libtenzir/builtins/contexts/bloom_filter.cpp

```diff
@@ -0,0 +1,240 @@
+      bloom_filter_.add(materialized_key);
+      key_values_list.emplace_back(std::move(materialized_key));
+    }
+    auto query_f = [key_values_list = std::move(key_values_list)](
+                     parameter_map params) -> caf::expected<expression> {
+      auto column = params["field"];
+      if (not column) {
+        return caf::make_error(ec::invalid_argument,
+                               "missing 'field' parameter");
+      }
+      return expression{
+        predicate{
+          field_extractor(*column),
+          relational_operator::in,
+          data{key_values_list},
+        },
+      };
+    };
+    return update_result{.update_info = show(),
+                         .make_query = std::move(query_f)};
+  }
+
+  auto update(chunk_ptr, context::parameter_map)
+    -> caf::expected<update_result> override {
+    // TODO: We're getting chunks piece-meal currently and therefore cannot
+    // determine the boundary of the stream. In theory, we should accumulate the
+    // chunks until the upstream pipeline is complete. What we need in the
+    // context API is a signal that we're done.
+    return ec::unimplemented;
+  }
+
+  auto update(context::parameter_map) -> caf::expected<update_result> override {
+    return caf::make_error(ec::unimplemented,
+                           "bloom-filter context can not be updated with void");
+  }
+
+  auto save() const -> caf::expected<chunk_ptr> override {
+    std::vector<std::byte> buffer;
+    if (auto err = convert(bloom_filter_, buffer))
+      return add_context(err, "failed to serialize Bloom filter context");
+    return chunk::make(std::move(buffer));
+  }
+
+private:
+  dcso_bloom_filter bloom_filter_;
+};
+
+class plugin : public virtual context_plugin {
+  auto initialize(const record&, const record&) -> caf::error override {
+    return caf::none;
+  }
+
+  auto name() const -> std::string override {
+    return "bloom-filter";
+  }
+
+  auto make_context(context::parameter_map parameters) const
+    -> caf::expected<std::unique_ptr<context>> override {
+    auto capacity = parameters["capacity"];
+    if (not capacity)
+      return caf::make_error(ec::parse_error, "no --capacity provided");
+    auto fpp = parameters["fp-prob"];
+    if (not fpp)
+      return caf::make_error(ec::parse_error, "no --fp-prob provided");
+    auto n = uint64_t{0};
+    if (not parsers::u64(*capacity, n)) {
+      return caf::make_error(ec::invalid_argument,
+                             "--capacity is not an integer");
+    }
+    auto p = double{0.0};
+    if (not parsers::real(*fpp, p)) {
+      return caf::make_error(ec::invalid_argument, "--fp-prob is not a double");
+    }
+    if (n == 0) {
+      return caf::make_error(ec::invalid_argument, "--capacity must be > 0");
+    }
+    if (p < 0.0 || p > 1.0) {
+      return caf::make_error(ec::invalid_argument, "--fp-prob not in (0,1)");
+    }
+    return std::make_unique<bloom_filter_context>(n, p);
```

Comment:
Please iterate over all key-value pairs in `parameters` so that you can return a meaningful error when an option is provided that may not be used with the bloom-filter context.

Reply:
Solved in [15e245a](https://github.com/tenzir/tenzir/pull/3834/commits/15e245a58d6294ece98c1e20b55e85366f72f0c7).